React Native Template App Code

mongomoe opened the pull request:
Note: Some Icons not loading in Android, and Android UI needs some cleaning


Video Demo (hosted on google drive): https://drive.google.com/file/d/1DhLHJcpwS0Uc1k7qGBQ1i7Izu0_7edH2/view?usp=sharing

cbush reviewed (approved):
Looks great! A few little comments.

mongomoe commented on react-native/src/Logout.js:
```diff
@@ -0,0 +1,41 @@
+  // logged in user
+  const signOut = () => {
+    if (user == null) {
+      console.warn("Not logged in, can't log out!");
```
The idea was taken from the RN tutorial - https://github.com/mongodb-university/realm-tutorial-react-native/blob/final/providers/AuthProvider.js#L82

I guess the idea, in that case, was to deal with the unique scenario in which a developer had imported this logout functionality into a view that users entered before log in.

mongomoe replied:
But even if we want to handle for that use case, I think this method might be better written as:
```
  const signOut = () => {
    if (user) {
      user.logOut();
    }

    navigation.popToTop()
  };
  ```

cbush replied:
Looks good.

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,182 @@
+          .sorted('summary');
+        // set the sorted Tasks to state as an array, so they can be rendered as a list
+        setTasks([...sortedTasks]);
+        // watch for changes to the Task collection, when tasks are created,
```
Fixed 👍

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,182 @@
+      if (realm) {
+        realm.close();
+        // set the references to null so the realm can't be used after it is closed
+        realm = null;
```
Good point, fixed now 👍

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,182 @@
+  // deleteTask() deletes a Task with a particular _id
+  const deleteTask = _id => {
+    const realm = realmReference.current;
+    const task = realm.objectForPrimaryKey('Task', _id); // search for a realm object with a primary key that is an objectId
```
Good idea, added null check on realm for this and all other methods that use the realm 👍

takameyer commented:
@mongomoe To fix the Android icons, add this line to the bottom of `android/app/build.gradle`
`apply from: "../../node_modules/react-native-vector-icons/fonts.gradle"`

takameyer commented on react-native/realm.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  appId: '<Your App ID>', // Replace with your App ID
+};
```
It might be helpful to throw an error here if the appId has not been set and link to the documentation.

cbush replied:
Actually I just noticed you can leave the actual app ID you're using here. The backend will automatically replace this file with your app ID. That said, it's still a good idea to make sure it is set somewhere. Also, I'll need to check with Andrew Chen that this format is ok as opposed to a loaded json file.

mongomoe replied:
Thanks for feedback guys. I've altered realm.js -> realm.json that looks like this:
```
{
  "appId": "<Your App ID>",
  "url": "http://localhost:8080"
}
```

and added:
```
if (!appId) {
  throw 'Missing Realm App ID. Set your App ID in realm.json';
}
```
where applicable

takameyer commented on react-native/src/App.js:
```diff
@@ -0,0 +1,49 @@
+              },
+            }}
+          />
+        </Stack.Navigator>
```
I recommend adding some logic here to see if the current user has been set and basing the `Navigator` contents off that.
For best practice, I would follow the recommendations in the documentation for [Authentication flows](https://reactnavigation.org/docs/auth-flow/)
This would help prevent the UI from going back to the `Welcome` screen without an explicit logout.

mongomoe replied:
Decided against this for this app, considering we only have two screens. A big part of the reason is to keep the app simple, by avoiding useReducer or some other state management library, without which we'd have to explicitly ignore warnings such as "Non-serializable values were found in the navigation state".

takameyer reviewed (approved):
All in all, I find the quality of the template quite good.  I have added a few suggestions that could help bring things into best practice.

What does need to be looked over is the README.md.  I think this could be improved and give some more explanation on how to setup sync, or at least link to the documentation.

Also if you want to fix icons in Android, add this line of code to `android/app/build.gradle`
`apply from: "../../node_modules/react-native-vector-icons/fonts.gradle"`

cbush commented on react-native/realm.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "appId": "<Your App ID>",
+  "url": "http://localhost:8080"
```
I think it should be "https://realm.mongodb.com", but it's probably ok to leave it off entirely for now. The support is there, so that's good.

mongomoe replied:
Sounds good, removed

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,190 @@
+  // state would.
+  const realmReference = useRef(null);
+
+  useEffect(() => {
```
I don't see the value in extracting this effect into its own hook. This effect is very specific to the Task View (in that it sets the Tasks to state and adds a listener on the Tasks. The realm itself is not stateful because it doesn't need to update the UI when it's changed.

 Originally I was thinking it could be a `useTasks` hook (sort of like in the tutorial that's currently in the docs - https://github.com/mongodb-university/realm-tutorial-react-native/blob/final/providers/TasksProvider.js, but I decided against it to keep the template app very simple.

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,190 @@
+        setTasks([]); // set the Tasks state to an empty array since the component is unmounting
+      }
+    };
+  }, [navigation]);
```
Good idea, done 👍

mongomoe commented on react-native/src/TasksView.js:
```diff
@@ -0,0 +1,190 @@
+      },
+    };
+
+    Realm.open(config)
```
Good idea, done 👍 

Changed the syntax slightly though to be a bit more scannable:
```
        const realm = realmReference.current;
        if (realm) { } 
```

mongomoe commented:
@takameyer  per your top level suggestions:

```To fix the Android icons, add this line to the bottom of android/app/build.gradle apply from: "../../node_modules/react-native-vector-icons/fonts.gradle" ```

I added that to the app/build.gradle and it works. Thanks for the tip. 

```
All in all, I find the quality of the template quite good. I have added a few suggestions that could help bring things into best practice.

What does need to be looked over is the README.md. I think this could be improved and give some more explanation on how to setup sync, or at least link to the documentation.
```

Thanks for the suggestions for best practices. I also created a follow up ticket for updating this README -> https://jira.mongodb.org/browse/DOCSP-15574
